src/coreclr/tools/Common/TypeSystem/Interop/MarshalAsDescriptor.cs

```diff
@@ -48,17 +48,24 @@ public enum NativeTypeKind : byte
 
     public class MarshalAsDescriptor
     {
+        private TypeDesc customMarshallerType;
+        private string cookie;
+
         public NativeTypeKind Type { get; }
         public NativeTypeKind ArraySubType { get; }
         public uint? SizeParamIndex { get; }
         public uint? SizeConst { get; }
+        public TypeDesc CustomMarshallerType => Type == NativeTypeKind.CustomMarshaler ? customMarshallerType : throw new InvalidOperationException("Custom marshaller type can be set only when using for CustomMarshaller");
+        public string Cookie => Type == NativeTypeKind.CustomMarshaler ? cookie : throw new InvalidOperationException("Cookie can be set only when using for CustomMarshaller");
 
-        public MarshalAsDescriptor(NativeTypeKind type, NativeTypeKind arraySubType, uint? sizeParamIndex, uint? sizeConst)
+        public MarshalAsDescriptor(NativeTypeKind type, NativeTypeKind arraySubType, uint? sizeParamIndex, uint? sizeConst, TypeDesc customMarshallerType, string cookie)
         {
             Type = type;
             ArraySubType = arraySubType;
             SizeParamIndex = sizeParamIndex;
             SizeConst = sizeConst;
+            this.customMarshallerType = customMarshallerType;
```

Comment:
No need to prefix with `this` if the field has underscore because that's the convention.